cpp/tests/test_abm.cpp

```diff
@@ -1481,3 +1420,148 @@ TEST(TestDiscreteDistribution, generate)
+
+    ScopedMockDistribution<testing::StrictMock<MockDistribution<mio::UniformDistribution<double>>>> mock_uniform_dist;
+    EXPECT_CALL(mock_uniform_dist.get_mock(), invoke)
+        .Times(testing::AtLeast(4))
+        .WillOnce(testing::Return(0.7))
+        .WillOnce(testing::Return(0.5))
+        .WillOnce(testing::Return(0.7))
+        .WillOnce(testing::Return(0.5));
+    ASSERT_EQ(testing_scheme.run_scheme(person1, loc_home), false);
+    ASSERT_EQ(testing_scheme.run_scheme(person2, loc_work), true);
+
+    ASSERT_EQ(person1.is_in_quarantine(), true);
+    ASSERT_EQ(person2.is_in_quarantine(), false);
+    ASSERT_EQ(person2.get_time_since_negative_test(), mio::abm::days(0));
+
+    testing_scheme.add_testing_rule(testing_rule1);
+    testing_scheme.remove_testing_rule(testing_rule1);
+    ASSERT_EQ(testing_scheme.run_scheme(person1, loc_home), true);
+}
+
+TEST(TestWorldTestingRule, testAddingAndUpdatingAndRunningTestingSchemes)
+{
+
+    auto world   = mio::abm::World();
+    auto home_id = world.add_location(mio::abm::LocationType::Home);
+    auto work_id = world.add_location(mio::abm::LocationType::Work);
+    auto person  = mio::abm::Person(home_id, mio::abm::InfectionState::Infected, mio::abm::AgeGroup::Age15to34,
+                                    world.get_global_infection_parameters());
+    auto& home   = world.get_individualized_location(home_id);
+    auto& work   = world.get_individualized_location(work_id);
+    person.set_assigned_location(home);
+    person.set_assigned_location(work);
+
+    auto testing_rule = mio::abm::TestingRule({}, {}, {});
+    testing_rule.add_infection_state(mio::abm::InfectionState::Infected);
+    testing_rule.add_infection_state(mio::abm::InfectionState::Carrier);
+    testing_rule.add_location_type(mio::abm::LocationType::Home);
+    testing_rule.add_location_type(mio::abm::LocationType::Work);
+
+    const auto testing_frequency = mio::abm::days(1);
+    const auto start_date        = mio::abm::TimePoint(20);
+    const auto end_date          = mio::abm::TimePoint(60 * 60 * 24 * 3);
+    const auto probability       = 1.0;
+    const auto test_type         = mio::abm::PCRTest();
+
+    auto testing_scheme =
+        mio::abm::TestingScheme({testing_rule}, testing_frequency, start_date, end_date, probability, test_type);
+
+    world.add_testing_scheme(testing_scheme);
+    auto current_time = mio::abm::TimePoint(0);
+    ASSERT_EQ(world.run_testing_schemes(person, work), true);
+    current_time = mio::abm::TimePoint(30);
+    world.update_testing_scheme_activity_status(current_time);
+    ScopedMockDistribution<testing::StrictMock<MockDistribution<mio::UniformDistribution<double>>>> mock_uniform_dist;
+    EXPECT_CALL(mock_uniform_dist.get_mock(), invoke)
+        .Times(testing::AtLeast(2))
+        .WillOnce(testing::Return(0.7))
+        .WillOnce(testing::Return(0.4));
+    ASSERT_EQ(world.run_testing_schemes(person, work), false);
```

Comment:
I do not understand this. If it returns 0.7 or 0.4, this is below the probability of 1.0 and the test should be executed, right?

Furthermore, there are many execution branches untested, e.g., if return larger probability, if test frequency says no test needed etc.